src/hotspot/share/prims/jvmtiAgentList.cpp

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "prims/jvmtiAgentList.hpp"
+
+#include "prims/jvmtiEnvBase.hpp"
+#include "prims/jvmtiExport.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/os.inline.hpp"
+#include "utilities/growableArray.hpp"
+
+JvmtiAgent* JvmtiAgentList::_list = nullptr;
+
+// Selection as a function of the filter.
+JvmtiAgent* JvmtiAgentList::Iterator::select(JvmtiAgent* agent) const {
+  while (agent != nullptr) {
+    if (_filter == ALL) {
+      return agent;
+    } else if (_filter == NOT_XRUN) {
+      if (!agent->is_xrun()) {
+        return agent;
+      }
+    } else if (_filter == JAVA) {
+      if (agent->is_jplis()) {
+        return agent;
+      }
+    } else if (_filter == NATIVE) {
+      if (!agent->is_jplis() && !agent->is_xrun()) {
+        return agent;
+      }
+    } else {
+      assert(_filter == XRUN, "invariant");
+      if (agent->is_xrun()) {
+        return agent;
+      }
+    }
+    agent = agent->next();
+  }
+  return nullptr;
+}
+
+static inline JvmtiAgent* head(JvmtiAgent** list) {
+  assert(list != nullptr, "invariant");
+  return Atomic::load_acquire(list);
+}
+
+
+// The storage list is a single cas-linked-list, to allow for concurrent iterations. Especially during initial loading of agents,
+// there exist an order requirement to iterate oldest -> newest. Our concurrent storage linked-list is newest -> oldest.
+// The correct order is preserved by the iterator, by storing a filtered set of entries in a stack.
+JvmtiAgentList::Iterator::Iterator(JvmtiAgent** list, Filter filter) : _stack(new GrowableArrayCHeap<JvmtiAgent*, mtServiceability>(16)), _filter(filter) {
```

Comment:
Nit: It'd be nice to make the lines 69-72 shorter.

Reply:
Ok. Fixed.